packages/controllers/src/snaps/Timer.ts

```diff
@@ -1,57 +1,128 @@
-export class Timer {
-  private _timeout?: NodeJS.Timeout | number;
-
-  private _start?: number;
+import { assert } from '../utils';
 
-  private _remaining: number;
+export type Status = 'stopped' | 'paused' | 'running' | 'finished';
 
-  private _callback?: () => void;
+export class Timer {
+  private state:
+    | { value: 'stopped'; remaining: number }
+    | {
+        value: 'paused';
+        remaining: number;
+        callback: () => void;
+      }
+    | {
+        value: 'running';
+        remaining: number;
+        callback: () => void;
+        start: number;
+        timeout?: unknown;
+      }
+    | { value: 'finished' };
 
+  /**
+   * If `ms` is smaller or equal to zero (including -Infinity), the callback is added to the event loop and executed async immediately
+   * If `ms` is +Infinity the timer never finishes.
+   *
+   * @throws {@link TypeError}. If `ms` is NaN.
+   * @param ms - The number of milliseconds before the callback is called after started.
+   */
   constructor(ms: number) {
-    this._remaining = ms;
-  }
-
-  isStarted() {
-    return Boolean(this._timeout && this._start);
+    assert(
+      !Number.isNaN(ms),
```

Comment:
If we want to add validation here we should also check that the number is positive.